swarm/swap_test.go

```diff
@@ -0,0 +1,468 @@
+				return errors.New("No swarm")
+			}
+			swarm := item.(*Swarm)
+			key, data, err := uploadFile(swarm)
+			if err != nil {
+				return err
+			}
+			log.Trace("file uploaded", "node", id, "key", key.String())
+			files = append(files, file{
+				addr:   key,
+				data:   data,
+				nodeID: id,
+			})
+		}
+
+		// File retrieval check is repeated until all uploaded files are retrieved from all nodes
+		// or until the timeout is reached.
+		for {
+			//we use a special retrieve function for swap which is optimized for parallel requests
+			//but does not leave many cascaded requests floating around
+			if retrieveForSwap(sim, files) == 0 {
+				break
+			}
+		}
+
+		//iterate all nodes
+		for _, node := range sim.NodeIDs() {
+			item, ok := sim.NodeItem(node, bucketKeySwarm)
+			if !ok {
+				log.Error("No swarm")
+				return errors.New("No swarm")
+			}
+			swarm := item.(*Swarm)
+
+			//submap for each node is a map of all nodes with the balance for that node
+			subBalances := make(map[enode.ID]int64)
+
+			//iterate all nodes again...
+			//get all balances with other peers for every node
+			for _, n := range sim.NodeIDs() {
+				if node == n {
+					continue
+				}
+
+				//get the peer's balance with this node
+				balance, err := swarm.swap.GetPeerBalance(n)
+				if err == nil {
+					subBalances[n] = balance
+					log.Debug(fmt.Sprintf("Balance of node %s to node %s: %d", node.TerminalString(), n.TerminalString(), balance))
+				} else {
+					log.Debug(fmt.Sprintf("Node %s has no balance with node %s", node.TerminalString(), n.TerminalString()))
+				}
+			}
+			//update the map for this node
+			balancesMap[node] = subBalances
+		}
+
+		return nil
+	})
+
+	sim.Close()
+	if result.Error != nil {
+		t.Fatal(result.Error)
+	}
+	//print all the balances if requested
+	if *printStats {
+		for k, v := range balancesMap {
+			fmt.Println(fmt.Sprintf("node %s balances:", k.TerminalString()))
+			for kk, vv := range v {
+				fmt.Println(fmt.Sprintf(".........with node %s: balance %d", kk.TerminalString(), vv))
+			}
+		}
+	}
+
+	//now iterate the whole map
+	//and check that every node k has the same
+	//balance with a peer as that peer with the node,
+	//but in inverted signs
+
+	//iterate the map
+	success := true
+	for k, mapForK := range balancesMap {
+		//iterate the submap
+		for n, balanceKwithN := range mapForK {
+			//iterate the main map again
+			mapForSubK := balancesMap[n]
+			log.Trace(fmt.Sprintf("balance of %s with %s: %d", k.TerminalString(), n.TerminalString(), balanceKwithN))
+			log.Trace(fmt.Sprintf("balance of %s with %s: %d", n.TerminalString(), k.TerminalString(), mapForSubK[k]))
+			//...check that they have the same balance in Abs terms and that it is not 0
+			if balanceKwithN+mapForSubK[k] != 0 && balanceKwithN != 0 {
+				log.Error(fmt.Sprintf("Expected balances to be a+b = 0 AND balance(a) != 0, but they are not, balance k with n:  %d, balance n with k: %d", balanceKwithN, mapForSubK[k]))
```

Comment:
you don't need `fmt.Sprintf`. just use `log.Error`

Reply:
Debatable. `log.Error` is fine, but if you want to construct more human-friendly log lines, sometimes `Sprintf` is convenient.

Reply:
i don't think you need so much text, you can describe the error in a comment and just `log.Error` the variable values. it will help to see the text for the first time when you dont _know_ the error, but once you do - you don't need the explanations anymore and just seeing the values would suffice

Reply:
Well I don't agree, as `fmt.Sprintf()` is for formatting and I wanted the error to be formatted the way I saw/see relevant.

But as you insist I will change that, because I want to move on with this PR and this is not something critical.

Reply:
as you please. not a blocker